File: src/main/python/com/hazza/LeetCodeByPython/SegmentOne/No6_ZigZag_Conversion.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/6 21:13
# https://leetcode.com/problems/zigzag-conversion/
import math


class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or len(s) <= numRows:
            return s
        i, step = 0, 1
        rows = [''] * numRows
        for c in s:
            rows[i] += c
            if i == 0:
                step = 1
            elif i == numRows - 1:
                step = -1
            i += step

        return ''.join(rows)


if __name__ == '__main__':
    solution = Solution()
    assert solution.convert('PAYPALISHIRING', 3) == 'PAHNAPLSIIGYIR'
    assert solution.convert('PAYPALISHIRING', 4) == 'PINALSIGYAHRPI'

File: src/main/java/com/hazza/leetcode/SegmentOne/No6ZigZagConversion.java
package com.hazza.leetcode.SegmentOne;

/**
 * https://leetcode.com/problems/zigzag-conversion/#/description
 */
public class No6ZigZagConversion {
    public String convert(String s, int numRows) {
        char[] c = s.toCharArray();
        int len = c.length;
        StringBuffer[] sb = new StringBuffer[numRows];
        for (int i = 0; i < numRows; ++i) sb[i] = new StringBuffer();
        for (int i = 0; i < len;) {
            for (int idx = 0; i < len && idx < numRows; ++idx) {
                sb[idx].append(c[i++]);
            }
            for (int idx = numRows - 2; i < len && idx >= 1; --idx) {
                sb[idx].append(c[i++]); //因为想要有zigzag的形状，至少要有三层，所以从倒数第二行开始
            }
        }
        for (int idx = 1; idx < numRows; ++idx) {
            sb[0].append(sb[idx]);
        }
        return sb[0].toString();
    }
}
